# sample generator for embedded questions
This notebook is for creating the random sample .txt files for the experiment.

In [61]:
import pandas as pd
import numpy as np

 # Contents
 1. [Constrain the data set to stimuli set](#Constrain-the-dataset-to-stimuli-set)
 2. [Figuring out the distribution of factors per list](#Figuring-out-the-distribution-of-factors-per-list)
 3. [Figure out how to collapse the matrix verb columns](#Figure-out-how-to-collapse-the-matrix-verb-columns)
 4. [Add in the paraphrases](#Add-in-the-paraphrases)
 5. [Split EntireSentence on Question](#Split-EntireSentence-on-Question)
 6. [Controls](#Controls)
 7. [Balancing factors](#Balancing-factors)
     1. [Modal Balancing](#Modal-Balancing)
     2. [Wh Balancing](#Wh-Balancing)
         1. [Who](#Who)
         2. [What](#What)
         3. [Where](#Where)
         4. [When](#When)
         5. [How](#How)
         6. [Why](#Why)
 8. [Generating-random-samples](#Generating-random-samples)
     1. [First Iteration](#First-Iteration)
     2. [Second Iteration](#Second-Itreation)
     3. [Third Iteration](#Third-Iteration)
     4. [Fourth Iteration](#Fourth-Iteration)
     5. [Fifth Iteration](#Fifth-Iteration)
     6. [Sixth Iteration](#Sixth-Iteration)
     7. [Final Set](#Final-Set)
 9. [Pilot Samples](#Pilot-Samples)

In [62]:
# import the database file from the TGrep2 searching
df = pd.read_csv("../results/swbd.tab", sep='\t', engine='python')

In [63]:
df.head()

Item_ID  \
0    3:43   
1   17:77   
2   21:45   
3   23:31   
4   96:22   

                                                                                                                                                                                                       Sentence  \
0                                                                 uh, first, um, i need *-1 to know, uh, how do you feel *t*-2 about, uh, about * sending, uh, an elderly, uh, family member to a nursing home?   
1                                                        and, uh, we were, i was fortunate in that i was personally acquainted with the, uh, people who, uh, *t*-1 ran the nursing home in our little hometown.   
2                                                                                          so, i was very comfortable, you know, in *-1 doing it when it got to the point that we had *-2 to do it *t*-3 *t*-4.   
3                          well, i had an occasion for my mother-in-law who *t*-1 had fell and needed * to be, you know, could not take care of herself anymore, was confined *-2 to a nursing home for a while   
4  i mean, for somebody who *t*-1 is, you know, for most of their life has, has, uh, not just merely had a farm but had ten children had a farm, ran everything because her husband was away in the coal mines.   

  HaveNeedTo Finite ModalPresent QuestionType EmbeddedSQ DegreeQ  \
0         no    yes           no     embedded        yes      no   
1         no    yes           no      subject        NaN      no   
2        yes    yes           no   embadjunct        NaN      no   
3        yes    NaN          yes     relative        NaN      no   
4         no    yes           no     relative        NaN      no   

  SubjectAuxInv WhAll  ... FullWhPhrase  \
0           yes   NaN  ...          NaN   
1           NaN   NaN  ...          who   
2           NaN   NaN  ...         when   
3           NaN   NaN  ...          who   
4           NaN   NaN  ...          who   

                                                                                                         JustMatrixClause  \
0  *-1 to know, uh, how do you feel *t*-2 about, uh, about * sending, uh, an elderly, uh, family member to a nursing home   
1                                                                                                                     NaN   
2                                              *-1 doing it when it got to the point that we had *-2 to do it *t*-3 *t*-4   
3                                                                                                                     NaN   
4                                                                                                                     NaN   

  DeterminerSubjPresent DeterminerNonSubjPresent WhNode     WhParse  \
0                    no                       no    NaN   (WRB how)   
1                    no                       no    NaN    (WP who)   
2                    no                       no    NaN  (WRB when)   
3                    no                       no    NaN    (WP who)   
4                    no                       no    NaN    (WP who)   

                                                                                                                                                                                 Question  \
0                                                                                   how do you feel *t*-2 about, uh, about * sending, uh, an elderly, uh, family member to a nursing home   
1                                                                                                                              who, uh, *t*-1 ran the nursing home in our little hometown   
2                                                                                                                           when it got to the point that we had *-2 to do it *t*-3 *t*-4   
3                                                 who 

In [64]:
# This makes the display show more info
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)

In [65]:
df.pivot_table(index=['QuestionType'], values="Question", aggfunc=len).groupby(["QuestionType"]).Question.transform(lambda x: x/len(df)).reset_index()

QuestionType  Question
0      adjunct  0.075203
1        cleft  0.064712
2   embadjunct  0.236886
3     embedded  0.160800
4     fragment  0.067654
5     relative  0.133150
6         root  0.124424
7      subject  0.137072

# Constrain the dataset to stimuli set
for experimental mock-up

First we have to remove the questions that we don;t want to include:
1. embedded questions only
2. no degree questions
3. no identity questions
4. generally only monomorphemic wh-phrases
5. only who-, what-, where-, when-, how-, and why-questions

In [66]:
critical = df[(df['QuestionType'] == 'embedded') # only embedded questions
              & 
              (df['Finite'] == 'no' )] # infinitival clauses
            

In [67]:
len(critical)

135

In [68]:
7*30

210

### Figuring out how many lists

135/20

6 lists of 20, 1 of 15

15 how / 6list
4 whats / 6 lists
then random sample 1 item from the remainders

last list is whatever's left over (total 15 items)


### Number of participants
7Lists*30 subjects/list = 210

# Combine contexts with constrained db

In [69]:
# read in df with contexts
cntxts = pd.read_csv("swbd_contexts.csv")

In [70]:
cntxts = cntxts.drop(columns="FollowingContext")

In [71]:
# get the indixes from critical
crit_index = critical.Item_ID

### Merge back in Wh colum
so we can sample proportionately based on Wh

In [72]:
df_Wh = critical[["Item_ID","Wh","Question"]].rename(columns={"Item_ID": "TGrepID"})

In [73]:
# subset to the items that are just the ones filtered in the previos section

# otherwise, if using the database file with contexts directly in there, then this step
# is not necessary
df_valid = cntxts[cntxts["TGrepID"].isin(set(crit_index))]

In [74]:
# Merge
df_valid = df_valid.merge(df_Wh, how = 'inner', indicator=False)

In [75]:
df_valid.head()

TGrepID  \
0  1049:139   
1   1130:56   
2   4531:36   
3   6397:23   
4   7981:19   

                                                                                                                                                                                                                         EntireSentence  \
0  i also thought about it, was of, uh, *-1 waiting *-2 to talk to you that, another thing that *t*-4 occurred to me is 0 there is not so much invasion of my privacy because i know how *-5 to behave *t*-6 such that there isn't *?*.   
1                                                                                                                                                       and if that's gone, um, i, i really don't know how *-1 to live *t*-2 very well,   
2                                                                                                                         and more and more people start *-2 believing them or wondering how * to combine them with other things *t*-1,   
3                                                                                                                                                   so, i really don't know how * to compare it to other big companies *t*-1, you know.   
4                                                                                                                                                                                    you don't even know who *-1 to payoff *t*-2, huh?.   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       PreceedingContext  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               ###speakera67.###yeah.###speakerb68.###so, maybe that is a, a little bit of what privacy is *t*-1.###speakera69.###yes.###exactly.###speakerb70.###speakera71.###uh-huh.   
1                                                                                                                                                                         ###i actually###for al-, of the time 0 i've spent *t*-1 there, i still don't quite understand how certain things that i assume 0 *t*-2 and require privacy and require not just that you be alone but actually that you have a sense of privacy *t*-3.###speakerb94.###yes.###speakera95.###because anyone can be alone for some period of time###but for me a lot of what i do *t*-1 requires a sense that there's this invisible barrier around me which people wil-, will respect *t*-2.###speakerb96.###yes.###speakera97.   
2  ###and, part of it is 0 california, you know, in, back in the sixties, had a lot of alternative movements###speakerb98.###speakera99.###and some of them fizzled out###and some of them were disastrous###and others of them, um, had an impact on the society around here.###and one of the ones that *t*-1 had an impact was, uh, people becoming interested in alternate practices,###i'm not sure if it was a meditation practice, or if it was, you know, 

# Split EntireSentence on Question 
This is necessary because we need to bold the question only and not the Matrix in the experimental file

In [76]:
# split EntireSentence at the string that matches the value in the 'question' column
df_valid["Matrix"] = df_valid.apply(lambda x: x['EntireSentence'].replace(x['Question'],"").strip(),axis=1)

In [77]:
# split that last punctuation off, to be added back on in .js script
df_valid["punct"] = df_valid["Matrix"].apply(lambda x: list(x)[-1])

In [78]:
# remove that final punct from the Matrix column
df_valid["Matrix"] = df_valid["Matrix"].apply(lambda x: x.replace(list(x)[-1], ' '))

In [79]:
df_valid.head()

TGrepID  \
0  1049:139   
1   1130:56   
2   4531:36   
3   6397:23   
4   7981:19   

                                                                                                                                                                                                                         EntireSentence  \
0  i also thought about it, was of, uh, *-1 waiting *-2 to talk to you that, another thing that *t*-4 occurred to me is 0 there is not so much invasion of my privacy because i know how *-5 to behave *t*-6 such that there isn't *?*.   
1                                                                                                                                                       and if that's gone, um, i, i really don't know how *-1 to live *t*-2 very well,   
2                                                                                                                         and more and more people start *-2 believing them or wondering how * to combine them with other things *t*-1,   
3                                                                                                                                                   so, i really don't know how * to compare it to other big companies *t*-1, you know.   
4                                                                                                                                                                                    you don't even know who *-1 to payoff *t*-2, huh?.   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       PreceedingContext  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               ###speakera67.###yeah.###speakerb68.###so, maybe that is a, a little bit of what privacy is *t*-1.###speakera69.###yes.###exactly.###speakerb70.###speakera71.###uh-huh.   
1                                                                                                                                                                         ###i actually###for al-, of the time 0 i've spent *t*-1 there, i still don't quite understand how certain things that i assume 0 *t*-2 and require privacy and require not just that you be alone but actually that you have a sense of privacy *t*-3.###speakerb94.###yes.###speakera95.###because anyone can be alone for some period of time###but for me a lot of what i do *t*-1 requires a sense that there's this invisible barrier around me which people wil-, will respect *t*-2.###speakerb96.###yes.###speakera97.   
2  ###and, part of it is 0 california, you know, in, back in the sixties, had a lot of alternative movements###speakerb98.###speakera99.###and some of them fizzled out###and some of them were disastrous###and others of them, um, had an impact on the society around here.###and one of the ones that *t*-1 had an impact was, uh, people becoming interested in alternate practices,###i'm not sure if it was a meditation practice, or if it was, you know, 

# Generating random samples

## First Iteration

6 lists of 20, 1 of 15

15 how / 6 list
4 whats / 6 lists
then random sample 1 item from the remainders

last list is whatever's left over (total 15 items)

In [80]:
for n in range(1,7):
    how_sample = df_valid[df_valid["Wh"] == "how"].sample(15)
    df_valid = df_valid.drop(how_sample.index)    
    
    what_sample = df_valid[df_valid["Wh"] == "what"].sample(4)
    df_valid = df_valid.drop(what_sample.index)
    
    final_sample = df_valid[
        (df_valid["Wh"] != "how") &
        (df_valid["Wh"] != "what")
    ].sample(1)
    df_valid = df_valid.drop(final_sample.index)

    
    total = pd.concat([how_sample,what_sample,final_sample])
    # save to file
    filename = f"../../experiments/clean_corpus/01_experiment/corpus_{n}.txt".format(n=n)
    total.to_csv(filename,header=True,sep="\t",index=False)
    
    
    
    

## Final Set

List = 15

10 how

In [81]:
len(df_valid)

15